### PR TITLE
fix(tls): Android wss:// CertificateBundleLoadFailure — merge CA trust across WSS + matchmaking (#24)

### LINKED ISSUE #24: Android WSS fails with CertificateBundleLoadFailure in release build
On Android release builds, Colyseus Godot native clients may fail to connect to `wss://` endpoints with:

```text
error 0: CertificateBundleLoadFailure
```

This happens before any room join succeeds. Multiple room/client instances fail with the same error, which suggests the failure is in TLS certificate bundle loading, not in room logic, auth, networking retries, or server protocol handling.

The issue appears device- or runtime-dependent: the same Android release build works for some users, but fails consistently for others.

Observed Behavior
In an Android release build using Godot 4.6.2-stable, all Colyseus rooms fail immediately after connecting:
```
status=connecting
status=error 0: CertificateBundleLoadFailure
```

The failure happens for multiple independent rooms using the same wss:// endpoint.

Diagnostics show:

```
os=Android
debug=false
endpoint=wss://example.com
last_error_code=0
last_error_message=CertificateBundleLoadFailure
connected=false
```

Expected Behavior
A Colyseus client using a valid public TLS endpoint should load CA roots and complete the WSS handshake, either by:

using Android system trust roots,
using the embedded Mozilla CA bundle,
or honoring network/tls/certificate_bundle_override.
Important Finding
The Android .so appears to include embedded Mozilla CA PEM data and strings such as:

```
Using bundled Mozilla CA certificates
network/tls/certificate_bundle_override
CertificateBundleLoadFailure
BEGIN CERTIFICATE
```

However, on affected Android release builds the client still fails with CertificateBundleLoadFailure.

Workaround Attempted
A client-side bootstrap was added before creating any Colyseus client:
```
const SOURCE_CERT_BUNDLE := "res://assets/certs/cacert.pem"
const USER_CERT_BUNDLE := "user://tls/cacert.pem"

DirAccess.make_dir_recursive_absolute("user://tls")

var source := FileAccess.open(SOURCE_CERT_BUNDLE, FileAccess.READ)
var data := source.get_buffer(source.get_length())

var target := FileAccess.open(USER_CERT_BUNDLE, FileAccess.WRITE)
target.store_buffer(data)

ProjectSettings.set_setting(
	"network/tls/certificate_bundle_override",
	USER_CERT_BUNDLE
)
```

The CA bundle is included in mobile exports:

include_filter="assets/certs/*.pem"
Diagnostics confirmed the file was copied successfully:
```
tls_override enabled=true configured=true path=user://tls/cacert.pem bytes=189462 error=
```

However, affected Android clients still failed with:
```
error 0: CertificateBundleLoadFailure
```

This suggests that either network/tls/certificate_bundle_override is not honored on Android in this path, user:// paths are not accepted by the native extension, or TLS/certificate initialization happens before the override is read.

Proposed Fix
In the native Godot Colyseus extension:

Ensure Android release builds can reliably load CA roots for WSS.
Prefer Android system trust store when available.
If using embedded Mozilla CA data, ensure it is available and parsed correctly on Android release builds.
Ensure network/tls/certificate_bundle_override is documented and works on Android, including what path formats are supported.
Consider logging the exact certificate source attempted and the low-level parse/open failure when CertificateBundleLoadFailure occurs.
Why This Matters
The failure occurs before Colyseus room logic runs, so frontend retry/reconnect logic cannot recover from it. The user only sees multiplayer unavailable, while the root cause is TLS initialization/certificate loading.

### PR DESCRIPTION
**Pull request by endel**
Fixes #24 — Android `wss://` release builds intermittently failed with a TLS certificate-bundle load failure (worked for some devices/users, failed for others against the same endpoint). Reported by @pierroo.

## Root cause
The SDK runs its **own mbedTLS** handshake (not Godot's TLS). It picked a **single** CA source in priority order — system store → bundled Mozilla → settings override — so on Android a device's system trust store could **shadow** the comprehensive bundled roots and abort TLS entirely if it lacked the server's root or failed to parse. That's the device-dependence. The override was also unreachable (the always-present bundled set sat ahead of it). Separately, **HTTPS matchmaking** uses a *different* trust stack (Zig `std.http`) that trusted only the OS store — same Android exposure, *before* the socket connects.

## Changes
- **Core (`websocket_transport.c`)** — merge all CA sources into one mbedTLS chain (bundled Mozilla always loaded as a device-independent baseline; system + override layered on top). Fail only if the chain is empty.
- **Core (`websocket_transport.c`)** — surface fatal TLS handshake failures: previously a cert-verification failure spun the state machine forever (silent hang); now closes with code `1015` + a specific reason.
- **Core (`http.zig`)** — HTTPS matchmaking now seeds `std.http.Client`'s trust store with system + bundled Mozilla + settings override, mirroring the WSS path (secure requests only; `http://` untouched).
- **Godot binding (`tls_certificates.c`)** — `network/tls/certificate_bundle_override` now honors `res://`/`user://` paths via `ProjectSettings.globalize_path()` (raw `fopen()` couldn't open Godot virtual paths — why the `user://` workaround silently fell back).

## Tests
- `tests/test_tls.zig` — drives the transport directly against a self-signed `wss://` echo server (no matchmaking): trusted-CA-via-settings → connects (the #24 regression), wrong/absent CA → fails verification, `tls_skip_verification` → connects. Wired into `tests.yml`.
- `platforms/godot/tests/test/test_wss.gd` — full `join_or_create` over `wss://` through a self-signed TLS proxy fronting `example-server`. Validated locally on **Godot 4.6.1 → 39/39 GUT pass**. Runs via `run-tests.sh` (generates certs + starts the proxy).
- Core unit tests: **65/65** (no regression). Extension cross-compiles for macOS/Linux/Windows (verified) and iOS/Android/Web targets.

## Notes
- Tagged `0.17.11-rc.1` (pre-release) so @pierroo can verify on the affected Android devices before a stable `0.17.11` — the device-level manifestation is the one thing we can't reproduce in CI.
- `std.http` doesn't verify IP-address SANs (only DNS/CN), so the Godot test connects via the `localhost` hostname; real deployments use hostnames, so this is an edge case.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/network/websocket_transport.c [hunk 1]
```diff
@@ -394,10 +401,17 @@ static void ws_tick_once(colyseus_transport_t* transport) {
...
     }
     else if (data->state == COLYSEUS_WS_TLS_HANDSHAKE) {
-        if (ws_tls_handshake_tick(data)) {
+        int hs = ws_tls_handshake_tick(data);
+        if (hs == WS_TLS_HS_DONE) {
             WS_LOG("TLS handshake complete");
             data->state = COLYSEUS_WS_HANDSHAKE_SENDING;
             ws_http_handshake_init(data);
+        } else if (hs == WS_TLS_HS_CERT_FAILED) {
+            ws_close_impl(transport, 1015, "TLS certificate verification failed");
```
Comment (bsharma-imperium) [inline src/network/websocket_transport.c:410]: RFC reserves 1015 for browser-internal TLS signaling i.e connection failures caused by a TLS handshake error. Shall we swap this to something from application-defined range (4000–4999) to be safe. Maybe 4015?

File: platforms/godot/src/tls_certificates.c [hunk 2]
```diff
@@ -24,35 +24,29 @@ void gdext_tls_certificates_set_api(GDExtensionInterfaceGetProcAddress p_get_pro
...
-        fprintf(stdout, "[Colyseus] Loading TLS certificates from: %s\n", override_path);
+        fprintf(stdout, "[Colyseus] Loading TLS certificate override from: %s\n", override_path);
         fflush(stdout);
-        
-        if (load_certificates_from_godot_path(override_path)) {
-            free(override_path);
-            return;
+
+        if (!load_certificates_from_godot_path(override_path)) {
+            fprintf(stderr, "[Colyseus] Failed to load certificate override; falling back to built-in CA roots\n");
```
Comment (bsharma-imperium) [inline platforms/godot/src/tls_certificates.c]: Maybe also add what next steps should be if we hit this error path, something like
```
"If using a packed res:// asset, copy it to user:// first."
```

File: src/network/websocket_transport.c [hunk 3]
```diff
@@ -937,9 +951,33 @@ static int tls_bio_recv(void* ctx, unsigned char* buf, size_t len) {
...
+    if (ret > 0) {
+        WS_LOG("CA source '%s' loaded (%d cert(s) skipped)", source, ret);
+    } else {
+        WS_LOG("CA source '%s' loaded", source);
+    }
+    return 1;
+}
+
+static bool ws_tls_init(colyseus_ws_transport_data_t* data, const char** out_err) {
     colyseus_tls_context_t* tls = malloc(sizeof(colyseus_tls_context_t));
```
Comment (bsharma-imperium) [inline src/network/websocket_transport.c:976]: Pre-existing nitpick: Should we switch to `calloc` instead? If `tls_skip_verification = true`, `ca_chain_initialized` would be garbage.

**Review by bsharma-imperium — APPROVED** [review]
LGTM. 
Added few comments.

File: platforms/godot/src/tls_certificates.c (hunk 2 above)
Comment (endel) [reply platforms/godot/src/tls_certificates.c]: Done in 0cb1f4c — the fallback log now ends with `(hint: copy a packed res:// asset to user:// first)`. Phrased as a hint rather than asserting res:// is the cause, since a res:// override works fine when running from the editor/source (it only fails for packed assets on an exported build).

File: src/network/websocket_transport.c (hunk 3 above)
Comment (endel) [reply src/network/websocket_transport.c:976]: `ca_chain_initialized` is actually set to `0` explicitly at websocket_transport.c:986 — unconditionally, before the `tls_skip_verify` branch and before `data->tls_ctx = tls` is published (which is the only way `ws_tls_cleanup` becomes reachable). So in the skip-verify path it's `0`, not garbage, and cleanup correctly skips `mbedtls_x509_crt_free`. `calloc` here would just redundantly zero the large mbedTLS sub-structs (ssl/conf/entropy/ctr_drbg) that get re-`init`'d right after, without fixing a real bug — so leaving it as `malloc`. Thanks for the careful read though!

File: src/network/websocket_transport.c (hunk 1 above)
Comment (endel) [reply src/network/websocket_transport.c:410]: Good point in general, but `1015` never reaches the wire on this path. `ws_close_impl` only queues a Close frame when `state == COLYSEUS_WS_CONNECTED` (websocket_transport.c:253), and a TLS handshake failure happens at `COLYSEUS_WS_TLS_HANDSHAKE` — before the HTTP upgrade and before any wslay context exists — so nothing is transmitted to the peer. It is purely a local code passed to `on_close`, which is exactly 1015's RFC-intended use, and it matches the SDK already using reserved `1006` as a local code here (including the sibling "TLS init failed"). A `4015` would land inside Colyseus's 4xxx protocol-code range (protocol.h: CONSENTED / SERVER_SHUTDOWN / MAY_TRY_RECONNECT / …) where it has no defined meaning, so it'd actually be less consistent. It's also reconnect-safe: not in `room_close_code_is_drop` (room.c:518), and the `!has_joined` guard means a TLS-stage close can't reach the reconnect path anyway. Keeping `1015`.

Comment (bsharma-imperium) [reply src/network/websocket_transport.c:410]: Cool. Thanks for the context.

File: src/network/websocket_transport.c (hunk 3 above)
Comment (bsharma-imperium) [reply src/network/websocket_transport.c:976]: Ohk. Its a old code, I didn't really look at the init, just `ca_chain_initialized` won't be set caught my mind. Lol.

**Comment by endel** [comment]
Thank you so much for the review @imperiumplay 🥳
